### PR TITLE
hwinfo: cleanup API doc and test

### LINKED ISSUE #22979: drivers: hwinfo: Build fails on some SoC
**Describe the bug**
The hwinfo driver does not build for:
 - nrf5340_dk_nrf5340_cpuappns
 - SAM3x siries
 - SAM4e and SAM4s series
 - STM32H7 series

**To Reproduce**
Build with one of those socks mentioned above

**Expected behavior**
Should build on every SoC that has a device-ID

**Impact**
annoyance. We want to declare the API as stabe, so it has to build for every supported SoC.

### PR DESCRIPTION
**Pull request by alexanderwachter**
Change the doxygen doc from "negative on error" to -ENOSUP only.
Enhance testing.
Fix build-errors on NRF53, STM32H7, SAM3x, SAM4e, SAM4s

Fixes  #22979

### CODE REVIEW

**Review by carlescufi — APPROVED** [review]

**Review by galak — CHANGES_REQUESTED** [review]
We should probably fix the tests/drivers/hwinfo/api/src/main.c to reflect what's documented here.

ie, -ENOTSUP is a valid return and shouldn't assert.  An other negative value should assert since that would be a violation of the API.

**Comment by alexanderwachter** [comment]
> We should probably fix the tests/drivers/hwinfo/api/src/main.c to reflect what's documented here.
> 
> ie, -ENOTSUP is a valid return and shouldn't assert. An other negative value should assert since that would be a violation of the API.

We only test on boards that have a hardware ID. It it returns -ENOSUP, we have a problem with the implementation.

**Comment by galak** [comment]
> We only test on boards that have a hardware ID. It it returns -ENOSUP, we have a problem with the implementation.

If you want to handle that in the test I think its fine, you can add to Kconfig option to tell if a driver is enabled or not and thus distinguish if -ENOSUP is a valid response or not.


```
diff --git a/drivers/hwinfo/Kconfig b/drivers/hwinfo/Kconfig
index 173dfa9672..7b5f0c0b51 100644
--- a/drivers/hwinfo/Kconfig
+++ b/drivers/hwinfo/Kconfig
@@ -16,9 +16,13 @@ config HWINFO_SHELL
        help
          Enable hwinfo Shell for testing.
 
+config HWINFO_HAS_DRIVER
+       bool
+
 config HWINFO_STM32
        bool "STM32 hwinfo"
        default y
+       select HWINFO_HAS_DRIVER
        depends on SOC_FAMILY_STM32
        select USE_STM32_LL_UTILS
        help
@@ -27,6 +31,7 @@ config HWINFO_STM32
 config HWINFO_NRF
        bool "NRF device ID"
        default y
+       select HWINFO_HAS_DRIVER
        depends on SOC_FAMILY_NRF
        help
          Enable Nordic NRF hwinfo driver.
...
```

**Comment by galak** [comment]
Than we can also get ride of 'hwinfo' from the board yaml and tests/drivers/hwinfo/api/testcase.yaml and change testcase.yaml filter on `CONFIG_HWINFO`

**Comment by galak** [comment]
I'm fine approving and merging the doc update PR, just be aware I'd like to see the other changes in testing for us to have this considered a stable API.

File: tests/drivers/hwinfo/api/testcase.yaml [hunk 1]
```diff
@@ -1,4 +1,3 @@
 tests:
   drivers.device_id:
     tags: driver
-    depends_on: hwinfo
```
Comment (galak) [inline tests/drivers/hwinfo/api/testcase.yaml]: make this `filter: CONFIG_HWINFO`

**Review by galak — CHANGES_REQUESTED** [review]
looks good, one minor tweak on the test yaml

we should also remove `hwinfo` from all board.yaml that have it.

**Review by galak — COMMENTED** [review]
On the 'Error returned:' in the test for negative retval, should we make that message a bit more verbose?

'Unexpected negative return value' or something?

**Comment by alexanderwachter** [comment]
> we should also remove `hwinfo` from all board.yaml that have it.

Why should we remove it? It states which boards have drivers for hwinfo.

**Comment by alexanderwachter** [comment]
> > we should also remove `hwinfo` from all board.yaml that have it.
> 
> Why should we remove it? It states which boards have drivers for hwinfo.

@galak I think I understand now. The yamls are for tests only and with the new tests, we also test for not supported. That way we can test on every board.

**Comment by alexanderwachter** [comment]
Somebody needs to test the SAM3X and SAM4X hwinfo. I don't have any of these boards.

**Review by galak — APPROVED** [review]

File: include/drivers/hwinfo.h [hunk 2]
```diff
@@ -39,7 +39,8 @@ extern "C" {
  * @param buffer  Buffer to write the ID to.
  * @param length  Max length of the buffer.
  *
- * @retval size of the device ID copied or negative on error.
+ * @retval size of the device ID copied
+ * @retval -ENOTSUP if there is no implementation for the particular device.
```
Comment (jfischer-phytec-iot) [inline include/drivers/hwinfo.h]: `size of the device ID copied or negative on error` is clear defined behavior, but not `-ENOTSUP if there is no implementation for the particular device`, what about other return values?

**Review by jfischer-no — DISMISSED** [review]

File: include/drivers/hwinfo.h (hunk 2 above)
Comment (alexanderwachter) [reply include/drivers/hwinfo.h]: There are no other errors.

Comment (jfischer-phytec-iot) [reply include/drivers/hwinfo.h]: > There are no other errors.

What is the reason to change it? Is it sure that future implementations would not need to return anything else as -ENOTSUP? What about an internal error like -EIO or -EBUSY?

Comment (alexanderwachter) [reply include/drivers/hwinfo.h]: I'd like to declare the API as stable, and the requirement from the API meeting was that I have to specify which errors could be returned. If we have another driver that could return another error, we have to announce this change, I guess.

Comment (alexanderwachter) [reply include/drivers/hwinfo.h]: This PR started as doc-change and escalated quickly to 8 commits and +128 lines.

Comment (jfischer-phytec-iot) [reply include/drivers/hwinfo.h]: I just do not see the reason to change clear defined `retval`.

Comment (alexanderwachter) [reply include/drivers/hwinfo.h]: I was not in the meeting. I don‘t know what was discussed there. @cales might know.

Comment (galak) [reply include/drivers/hwinfo.h]: From dev-review meeting: Document the specific error for -ENOTSUP and than document that any other negative value would be a driver specific error.

**Comment by nandojve** [comment]
> Somebody needs to test the SAM3X and SAM4X hwinfo. I don't have any of these boards.
> @fallrisk @mnkp @nandojve

I help with tests on SAM4S/E variants. I don't have any SAM3 board.

**Comment by ioannisg** [comment]
> > Somebody needs to test the SAM3X and SAM4X hwinfo. I don't have any of these boards.
> > @fallrisk @mnkp @nandojve
> 
> I help with tests on SAM4S/E variants. I don't have any SAM3 board.

I can test on SAM3X, hold on

**Comment by alexanderwachter** [comment]
@nandojve @ioannisg can you guys check if the ID differs on two different boards? You can use the shell module to check. Just to make sure that I got the address right.

**Comment by nandojve** [comment]
> @nandojve @ioannisg can you guys check if the ID differs on two different boards? You can use the shell module to check. Just to make sure that I got the address right.

Hi @alexanderwachter, I already tested on both boards. I separated values for better reading on the post. I added values that I get using Atmel Studio.
```
SAM4E
uart:~$ hwinfo devid 
Length: 16
ID: 0x00313353_50334337_30333032_30303436
[0] 0x53333100
[1] 0x37433350
[2] 0x32303330
[3] 0x36343030
uart:~$ 

SAM4S
uart:~$ hwinfo devid
Length: 16
ID: 0x00310053_594a5135_30323032_33303830
[0] 0x53003100
[1] 0x35514a59
[2] 0x32303230
[3] 0x30383033
uart:~$ 
```
Below I executed a demo inside Atmel Studio to get Device ID.
```
-- Flash Read Unique Identifier Example --
-- SAM4S-XPLAINED --
-- Compiled: Feb 20 2020 18:26:05 --
-I- Reading 128 bits Unique Identifier
	    [0]		[1]	  [2]	     [3]
-I- ID: 1392521472, 894519897, 842019376, 808988723
SAM4S
	0x53003100, 0x35514a59, 0x32303230, 0x30383033
SAM4E
        0x53333100, 0x37433350, 0x32303330, 0x36343030
```

**Comment by alexanderwachter** [comment]
Thanks a lot @nandojve !

File: drivers/hwinfo/Kconfig [hunk 3]
```diff
@@ -21,55 +24,63 @@ config HWINFO_STM32
 	default y
 	depends on SOC_FAMILY_STM32
 	select USE_STM32_LL_UTILS
```
Comment (erwango) [inline drivers/hwinfo/Kconfig]: This could actually be removed since driver only require headers, not utils functions

**Review by erwango — CHANGES_REQUESTED** [review]
One minor related change requested, otherwise I'm fine with the change

File: drivers/hwinfo/Kconfig (hunk 3 above)
Comment (alexanderwachter) [reply drivers/hwinfo/Kconfig]: Thanks. Removed.

**Comment by erwango** [comment]
Before approving: I've tried to run sanitycheck, and there's a strange ERROR message reported (with no impact on final result):
```
$ sanitycheck -T tests/drivers/hwinfo/ -p nucleo_f401re
Renaming output directory to /local/mcu/zephyrproject/zephyr/sanity-out.2
INFO    - JOBS: 8
ERROR   - /local/mcu/zephyrproject/zephyr/tests/drivers/hwinfo/api/src/main.c: found invalid #endif, #ifdef in ztest_test_suite()
INFO    - Building initial testcase list...
INFO    - 1 test configurations selected, 0 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Total complete:    1/   1  100%  skipped:    0, failed:    0
INFO    - 1 of 1 tests passed (100.00%), 0 failed, 0 skipped with 0 warnings in 4.94 seconds
INFO    - In total 1 test cases were executed on 1 out of total 229 platforms (0.44%)
```

Can you check the ERROR line ?

**Comment by alexanderwachter** [comment]
@erwango pushed a new version that does not have the error above.
The fix is not to have a #ifdef in the test-suite but replace the tests with dummies depending on the configuration.

**Review by erwango — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
@jhedberg this should be ready

**Comment by carlescufi** [comment]
@jhedberg this should be ready

**Comment by jhedberg** [comment]
> @jhedberg this should be ready

@carlescufi ok, I'll merge once CI completes.

**Review by jhedberg — APPROVED** [review]

**Comment by alexanderwachter** [comment]
@jhedberg the test are timing out.  tests/drivers/hwinfo/api/drivers.device_id tests are all green.

**Comment by jhedberg** [comment]
> @jhedberg the test are timing out. tests/drivers/hwinfo/api/drivers.device_id tests are all green.

Yes, I've had to restart Shippable a few times because of this. Fingers crossed that it wont time out again.

**Pull request merged**